cloudinit/sources/DataSourceVultr.py

```diff
@@ -0,0 +1,153 @@
+        BUILTIN_DS_CONFIG['retries'] = self.ds_cfg.get(
+            'retries', BUILTIN_DS_CONFIG['retries'])
+        BUILTIN_DS_CONFIG['timeout'] = self.ds_cfg.get(
+            'timeout', BUILTIN_DS_CONFIG['timeout'])
+        BUILTIN_DS_CONFIG['wait'] = self.ds_cfg.get(
+            'wait', BUILTIN_DS_CONFIG['wait'])
+
+    # Initiate data and check if Vultr
+    def _get_data(self):
+        LOG.debug("Detecting if machine is a Vultr instance")
+        if not vultr.is_vultr():
+            LOG.debug("Machine is not a Vultr instance")
+            return False
+
+        LOG.debug("Machine is a Vultr instance")
+
+        # Fetch metadata
+        md = self.get_metadata()
+
+        config = vultr.generate_config(md)
+
+        # This requires info generated in the vendor config
+        script = md['startup-script']
+        user_scripts = []
+        user_scripts = vultr.generate_user_scripts(script, config)
+
+        # Dump vendor config so diagnosing failures is manageable
+        LOG.debug("Vultr Vendor Config:")
+        LOG.debug(json.dumps(config))
+
+        self.metadata_full = md
+        self.metadata['instanceid'] = self.metadata_full['instanceid']
+        self.metadata['local-hostname'] = self.metadata_full['hostname']
+
+        # Default hostname is "vultr"
+        if self.metadata['local-hostname'] == "":
+            self.metadata['local-hostname'] = "vultr"
+
+        self.metadata['public-keys'] = md["public-keys"].splitlines()
+        self.userdata_raw = md["user-data"]
+        if self.userdata_raw == "":
+            self.userdata_raw = None
+
+        self.vendordata_raw = []
+        for uscript in user_scripts:
+            self.vendordata_raw.append(uscript)
+        self.vendordata_raw.append("#cloud-config\n%s" % json.dumps(config))
+
+        # Dump some data so diagnosing failures is manageable
+        LOG.debug("SUBID: %s", self.metadata['instanceid'])
+        LOG.debug("Hostname: %s", self.metadata['local-hostname'])
+        if self.userdata_raw is not None:
+            LOG.debug("User-Data:")
+            LOG.debug(self.userdata_raw)
+
+        return True
+
+    # Get the metadata by flag
+    def get_metadata(self):
+        return vultr.get_metadata(self.ds_cfg['url'],
+                                  self.ds_cfg['timeout'],
+                                  self.ds_cfg['retries'],
+                                  self.ds_cfg['wait'])
+
+    # Compare subid as instance id
+    def check_instance_id(self, sys_cfg):
+        if not vultr.is_vultr():
+            return False
+
+        # Baremetal has no way to implement this in local
+        if vultr.is_baremetal():
+            return False
+
+        subid = vultr.get_sysinfo()['subid']
+        return sources.instance_id_matches_system_uuid(subid)
+
+    # Currently unsupported
+    @property
+    def launch_index(self):
+        return None
+
+    @property
+    def network_config(self):
+        md = self.get_metadata()
```

Comment:
**need to change**

_get_metadata is guaranteed to have been called when network_config is called.

no need to re-call get_metadata()  you can see other datasource examples, they mainly use the cached value and then call a converter functio to turn it into what cloud-init uses.

Reply:
This is cached via lrucache. Did you want me to cache it further?